### PR TITLE
Rework SESSION_GLOBALS API

### LINKED ISSUE #84954: Don't allow overwriting `SESSION_GLOBALS`
Since `rustc_span::SESSION_GLOBALS` uses `scoped-tls`, it can be overwritten by calling `SESSION_GLOBALS.set` from within the closure passed to the original call to `SESSION_GLOBALS.set`. Any `Span`s created in the 'outer' `SessionGlobals` may become unusable within the inner `SessionGlobals`, since they will be pointing to the wrong interner. See https://github.com/rust-lang/rust/pull/84953 for an example of this happening.

There is no reason to modify `SESSION_GLOBALs` once it's set for the first time. We should make it private to the `rustc_span` crate, and only expose setters which assert that it is currently unset.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #84954.

<s>Needs #84953 to be merged first (I cherry-picked its commits to have CI pass).</s> (done)

r? @Aaron1011

### CODE REVIEW

File: compiler/rustc_span/src/lib.rs [hunk 1]
```diff
@@ -100,19 +100,35 @@ impl SessionGlobals {
...
+pub fn create_session_globals_then<R>(edition: Edition, f: impl FnOnce() -> R) -> R {
     let session_globals = SessionGlobals::new(edition);
+    assert!(!SESSION_GLOBALS.is_set(), "SESSION_GLOBALS should never be overwritten!");
     SESSION_GLOBALS.set(&session_globals, f)
 }
 
-pub fn with_default_session_globals<R>(f: impl FnOnce() -> R) -> R {
-    with_session_globals(edition::DEFAULT_EDITION, f)
+pub fn set_session_globals_then<R>(session_globals: &SessionGlobals, f: impl FnOnce() -> R) -> R {
+    #[cfg(not(parallel_compiler))]
```
Comment (Aaron1011) [inline compiler/rustc_span/src/lib.rs:110]: Can you add a FIXME explaining why this is necessary?

Comment (GuillaumeGomez) [reply compiler/rustc_span/src/lib.rs:110]: Sure.

**Comment by GuillaumeGomez** [comment]
Funny bug in the compiler error output (`use rustc_span::symbol::Ident;` is only present once):

```
Diff in /checkout/compiler/rustc_ast_pretty/src/pprust/tests.rs at line 1:
 use super::*;
 use rustc_ast as ast;
-use rustc_span::symbol::Ident;
-use rustc_span::symbol::Ident;
```

**Comment by GuillaumeGomez** [comment]
I get a timeout on `src/test/ui/tcp-stress.rs`. If someone has an idea on what is going on?

**Comment by GuillaumeGomez** [comment]
Now come the interesting part: I need to update `rustfmt` to make the CI happy, but I need this PR to be merged in order to not break rustfmt compilation. How does it work in such case?

**Comment by GuillaumeGomez** [comment]
Or I can keep the current function used in rustfmt, wait for this PR to be merged but add a deprecated attribute and then remove it once rustfmt has been updated. I like how it sounds. :)

**Comment by Aaron1011** [comment]
@GuillaumeGomez https://github.com/rust-lang/rust/pull/82208 has been approved - once it gets merged, you should be able to update the `rustfmt` API usages within this PR.

**Comment by GuillaumeGomez** [comment]
Subtree is now merged so I'll rebase and make the whole migration in one shot.

**Comment by GuillaumeGomez** [comment]
I need some help to try to understand what's going on with rustfmt.

**Comment by bstrie** [comment]
Triage here, any progress? Ordinarily I would reset the status to waiting-on-author, but I see that you manually set it to waiting-on-review.

**Comment by GuillaumeGomez** [comment]
The PR is now ready for review! \o/

**Comment by GuillaumeGomez** [comment]
@Aaron1011 doesn't seem to be around so let's assign someone else.

r? @pnkfelix

**Comment by GuillaumeGomez** [comment]
Or anyone else from the @rust-lang/compiler team?

**Comment by pnkfelix** [comment]
This seems fine.

The most important thing, AFAICT, is the new assertions that have been added to ensure that the session globals are not newly created (i.e. overwritten) once they have been established on a given thread.

**Pull request merged**